modules/pricer/src/main/java/com/opengamma/strata/pricer/rate/ImmutableRatesProviderBuilder.java

```diff
@@ -0,0 +1,356 @@
+   * <p>
+   * This adds the specified forward curve to the provider.
+   * This operates using {@link Map#put(Object, Object)} semantics using the index as the key.
+   * 
+   * @param index  the index of the curve
+   * @param forwardCurve  the Overnight index forward curve
+   * @return this, for chaining
+   */
+  public ImmutableRatesProviderBuilder overnightIndexCurve(OvernightIndex index, Curve forwardCurve) {
+    ArgChecker.notNull(index, "currency");
+    ArgChecker.notNull(forwardCurve, "forwardCurve");
+    this.indexCurves.put(index, forwardCurve);
+    return this;
+  }
+
+  /**
+   * Adds a index forward curve to the provider with associated time-series.
+   * <p>
+   * This adds the specified forward curve and time-series to the provider.
+   * This operates using {@link Map#put(Object, Object)} semantics using the index as the key.
+   * 
+   * @param index  the index of the curve
+   * @param forwardCurve  the index forward curve
+   * @param timeSeries  the associated time-series
+   * @return this, for chaining
+   */
+  public ImmutableRatesProviderBuilder overnightIndexCurve(
+      OvernightIndex index,
+      Curve forwardCurve,
+      LocalDateDoubleTimeSeries timeSeries) {
+
+    ArgChecker.notNull(index, "currency");
+    ArgChecker.notNull(forwardCurve, "forwardCurve");
+    ArgChecker.notNull(timeSeries, "timeSeries");
+    this.indexCurves.put(index, forwardCurve);
+    this.timeSeries.put(index, timeSeries);
+    return this;
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Adds index forward curves to the provider.
+   * <p>
+   * This adds the specified index forward curves to the provider.
+   * This operates using {@link Map#putAll(Map)} semantics using the index as the key.
+   * 
+   * @param indexCurves  the index forward curves
+   * @return this, for chaining
+   */
+  public ImmutableRatesProviderBuilder indexCurves(Map<Index, Curve> indexCurves) {
+    ArgChecker.noNulls(indexCurves, "indexCurves");
+    for (Entry<Index, Curve> entry : indexCurves.entrySet()) {
+      Index index = entry.getKey();
+      if (index instanceof IborIndex) {
+        iborIndexCurve((IborIndex) index, entry.getValue());
+      } else if (index instanceof OvernightIndex) {
+        overnightIndexCurve((OvernightIndex) index, entry.getValue());
+      } else {
+        throw new IllegalArgumentException("Unknown index type: " + index);
+      }
+    }
+    return this;
+  }
+
+  /**
+   * Adds index forward curves to the provider with associated time-series.
+   * <p>
+   * This adds the specified index forward curves to the provider.
+   * This operates using {@link Map#putAll(Map)} semantics using the index as the key.
+   * 
+   * @param indexCurves  the index forward curves
+   * @param timeSeries  the associated time-series
+   * @return this, for chaining
+   */
+  public ImmutableRatesProviderBuilder indexCurves(
+      Map<Index, Curve> indexCurves,
```

Comment:
Wildcards for the `Index` and `Curve` type parameters would be kinder to callers.